README.md
script.library.info.update
==========================

Updates your XBMC library with new info

changelog.txt
[B]0.0.3[/B]
- Added to XBMC pre-Frodo repo

[B]0.0.1[/B]
- Initial test version

resources/lib/utils.py
#import modules
import socket
import xbmc
import xbmcgui
import unicodedata
import urllib2
import sys

### get addon info
__addon__       = ( sys.modules[ "__main__" ].__addon__ )
__addonname__   = ( sys.modules[ "__main__" ].__addonname__ )
__icon__        = ( sys.modules[ "__main__" ].__icon__ )
__localize__    = ( sys.modules[ "__main__" ].__localize__ )
__addonprofile__= ( sys.modules[ "__main__" ].__addonprofile__ )


### import libraries
from urllib2 import HTTPError, URLError
from resources.lib.script_exceptions import *

# Use json instead of simplejson when python v2.7 or greater
if sys.version_info < (2, 7):
    import json as simplejson
else:
    import simplejson
# Commoncache plugin import
try:
    import StorageServer
except:
    import storageserverdummy as StorageServer

cache = StorageServer.StorageServer("LibraryInfoUpdate",48)

### Adjust default timeout to stop script hanging
socket.setdefaulttimeout(20)
### Cache bool
CACHE_ON = True

### Declare dialog
dialog = xbmcgui.DialogProgress()


# Fixes unicode problems
def string_unicode(text, encoding='utf-8'):
    try:
        text = unicode( text, encoding )
    except:
        pass
    return text

def normalize_string(text):
    try:
        text = unicodedata.normalize('NFKD', string_unicode(text)).encode('ascii', 'ignore')
    except:
        pass
    return text

# Define log messages
def log(txt, severity=xbmc.LOGDEBUG):
    if severity == xbmc.LOGDEBUG and not __addon__.getSetting("debug_enabled") == 'true':
        pass
    else:
        try:
            message = ('%s: %s' % (__addonname__,txt) )
            xbmc.log(msg=message, level=severity)
        except UnicodeEncodeError:
            try:
                message = normalize_string('%s: %s' % (__addonname__,txt) )
                xbmc.log(msg=message, level=severity)
            except:
                message = ('%s: UnicodeEncodeError' %__addonname__)
                xbmc.log(msg=message, level=xbmc.LOGWARNING)

# Define dialogs
def dialog_msg(action,
               percentage = 0,
               line0 = '',
               line1 = '',
               line2 = '',
               line3 = '',
               background = False,
               nolabel = __localize__(32026),
               yeslabel = __localize__(32025)):
    # Fix possible unicode errors 
    line0 = line0.encode( 'utf-8', 'ignore' )
    line1 = line1.encode( 'utf-8', 'ignore' )
    line2 = line2.encode( 'utf-8', 'ignore' )
    line3 = line3.encode( 'utf-8', 'ignore' )

    # Dialog logic
    if not line0 == '':
        line0 = __addonname__ + line0
    else:
        line0 = __addonname__
    if not background:
        if action == 'create':
            dialog.create(__addonname__, line1, line2, line3)
        if action == 'update':
            dialog.update(percentage, line1, line2, line3)
        if action == 'close':
            dialog.close()
        if action == 'iscanceled':
            if dialog.iscanceled():
                return True
            else:
                return False
        if action == 'okdialog':
            xbmcgui.Dialog().ok(line0, line1, line2, line3)
        if action == 'yesno':
            return xbmcgui.Dialog().yesno(line0, line1, line2, line3, nolabel, yeslabel)
    if background:
        if (action == 'create' or action == 'okdialog'):
            if line2 == '':
                msg = line1
            else:
                msg = line1 + ': ' + line2
            xbmc.executebuiltin("XBMC.Notification(%s, %s, 7500, %s)" % (line0, msg, __icon__))

# Retrieve JSON data from cache function
def get_data(url, data_type ='json'):
    log('API: %s'% url)
    if CACHE_ON:
        result = cache.cacheFunction(get_data_new, url, data_type)
    else:
        result = get_data_new(url, data_type)    
    if not result:
        result = 'Empty'
    return result

# Retrieve JSON data from site
def get_data_new(url, data_type):
    log('Cache expired. Retrieving new data')
    data = []
    try:
        request = urllib2.Request(url)
        # TMDB needs a header to be able to read the data
        if url.startswith("http://api.themoviedb.org"):
            request.add_header("Accept", "application/json")
        # Add some delay to stop trashing the fanart.tv server
        elif url.startswith("http://fanart.tv/"):
            xbmc.sleep(1000)
        req = urllib2.urlopen(request)
        if data_type == 'json':
            data = simplejson.loads(req.read())
            if not data:
                data = 'Empty'
        else:
            data = req.read()
        req.close()
    except HTTPError, e:
        if e.code == 400:
            raise HTTP400Error(url)
        elif e.code == 404:
            raise HTTP404Error(url)
        elif e.code == 503:
            raise HTTP503Error(url)
        else:
            raise DownloadError(str(e))
    except URLError:
        raise HTTPTimeout(url)
    except socket.timeout, e:
        raise HTTPTimeout(url)
    except:
        data = 'Empty'
    return data

# Clean filenames for illegal character in the safest way for windows
def clean_filename(filename):
    illegal_char = '<>:"/\|?*'
    for char in illegal_char:
        filename = filename.replace( char , '' )
    return filename

def save_nfo_file(data, target):
    try:
        # open source path for writing and write xmlSource
        file_object = open(target.encode("utf-8"), "w")
        file_object.write(data.encode( "utf-8" ))
        file_object.close()
        return True
    except Exception, e:
        log(str(e), xbmc.LOGERROR)
        return False

resources/lib/fileops.py
#import modules
import os
import socket
import urllib2
import urllib
import xbmc
import xbmcvfs

### import libraries
from traceback import print_exc
from urllib2 import HTTPError, URLError
from resources.lib.script_exceptions import *
from resources.lib import utils
from resources.lib.utils import log

THUMBS_CACHE_PATH = xbmc.translatePath( "special://profile/Thumbnails/Video" )


### adjust default timeout to stop script hanging
timeout = 10
socket.setdefaulttimeout(timeout)

class fileops:
    """
    This class handles all types of file operations needed by
    script.extrafanartdownloader (creating directories, downloading
    files, copying files etc.)
    """

    def __init__(self):
        log("Setting up fileops")
        self._exists = lambda path: xbmcvfs.exists(path)
        self._rmdir = lambda path: xbmcvfs.rmdir(path)
        self._mkdir = lambda path: xbmcvfs.mkdir(path)
        self._delete = lambda path: xbmcvfs.delete(path)

        self.downloadcount = 0
        self.tempdir = os.path.join(utils.__addonprofile__, 'temp')
        if not self._exists(self.tempdir):
            if not self._exists(utils.__addonprofile__):
                if not self._mkdir(utils.__addonprofile__):
                    raise CreateDirectoryError(utils.__addonprofile__)
            if not self._mkdir(self.tempdir):
                raise CreateDirectoryError(self.tempdir)
        
    def _copy(self, source, target):
        return xbmcvfs.copy(source.encode("utf-8"), target.encode("utf-8"))

    ### Delete file from all targetdirs
    def _delete_file_in_dirs(self, filename, targetdirs, reason, media_name = '' ):
        isdeleted = False
        for targetdir in targetdirs:
            path = os.path.join(targetdir, filename)
            if self._exists(path):
                self._delete(path)
                log("[%s] Deleted (%s): %s" % (media_name, reason, path))
                isdeleted = True
        if not isdeleted:
            log("[%s] Ignoring (%s): %s" % (media_name, reason, filename))


    # copy file from temp to final location
    def _copyfile(self, sourcepath, targetpath, media_name = ''):
        targetdir = os.path.dirname(targetpath).encode("utf-8")
        if not self._exists(targetdir):
            if not self._mkdir(targetdir):
                raise CreateDirectoryError(targetdir)
        if not self._copy(sourcepath, targetpath):
            raise CopyError(targetpath)
        else:
            log("[%s] Copied successfully: %s" % (media_name, targetpath) )

    # download file
    def _downloadfile(self, url, filename, targetdir, media_name):
        try:
            temppath = os.path.join(self.tempdir, filename)
            tempfile = open(temppath, "wb")
            response = urllib2.urlopen(url)
            tempfile.write(response.read())
            tempfile.close()
            response.close()
        except HTTPError, e:
            if e.code == 404:
                raise HTTP404Error(url)
            else:
                raise DownloadError(str(e))
        except URLError:
            raise HTTPTimeout(url)
        except socket.timeout, e:
            raise HTTPTimeout(url)
        except Exception, e:
            log(str(e), xbmc.LOGNOTICE)
        else:
            log("[%s] Downloaded: %s" % (media_name, filename))
            self.downloadcount += 1
            targetpath = os.path.join(targetdir, filename)
            self._copyfile(temppath, targetpath, media_name)

        
        def cleanup(self):
            if self.fileops._exists(self.fileops.tempdir):
                dialog_msg('update', percentage = 100, line1 = __localize__(32005), background = False)
                log('Cleaning up temp files')
                for x in os.listdir(self.fileops.tempdir):
                    tempfile = os.path.join(self.fileops.tempdir, x)
                    self.fileops._delete(tempfile)
                    if self.fileops._exists(tempfile):
                        log('Error deleting temp file: %s' % tempfile, xbmc.LOGERROR)
                self.fileops._rmdir(self.fileops.tempdir)
                if self.fileops._exists(self.fileops.tempdir):
                    log('Error deleting temp directory: %s' % self.fileops.tempdir, xbmc.LOGERROR)
                else:
                    log('Deleted temp directory: %s' % self.fileops.tempdir)

resources/lib/media_setup.py
#import modules
import os
import xbmc
import urllib
import sys

# Use json instead of simplejson when python v2.7 or greater
if sys.version_info < (2, 7):
    import json as simplejson
else:
    import simplejson

### import libraries
from resources.lib.utils import log

def __get_library(media_type):
    log('Using JSON for retrieving %s info' %media_type)
    Medialist = []
    if media_type == 'tvshow':
        json_response = xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "VideoLibrary.GetTVShows", "params": {"properties": ["file", "imdbnumber"], "sort": { "method": "label" } }, "id": 1}')
        json_response = unicode(json_response, 'utf-8', errors='ignore')
        jsonobject = simplejson.loads(json_response)
        if jsonobject['result'].has_key('tvshows'):
            for item in jsonobject['result']['tvshows']:
                # Search for season information
                json_response_season = xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "VideoLibrary.GetSeasons", "params": {"properties": ["season"], "sort": { "method": "label" }, "tvshowid":%s }, "id": 1}' %item.get('tvshowid',''))
                jsonobject_season = simplejson.loads(json_response_season)
                # Get start/end and total seasons
                if jsonobject_season['result'].has_key('limits'):
                    season_limit = jsonobject_season['result']['limits']
                # Get the season numbers
                seasons_list =[]
                if jsonobject_season['result'].has_key('seasons'):
                    seasons = jsonobject_season['result']['seasons']
                    for season in seasons:
                        seasons_list.append(season.get('season')) 
                Medialist.append({'id': item.get('imdbnumber',''),
                                  'tvshowid': item.get('tvshowid',''),
                                  'name': item.get('label',''),
                                  'path': media_path(item.get('file','')),
                                  'seasontotal': season_limit.get('total',''),
                                  'seasonstart': season_limit.get('start',''),
                                  'seasonend': season_limit.get('end',''),
                                  'seasons': seasons_list})

    elif media_type == 'movie':
        json_response = xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "VideoLibrary.GetMovies", "params": {"properties": ["file", "imdbnumber", "year", "trailer", "streamdetails", "votes", "rating", "mpaa"], "sort": { "method": "label" } }, "id": 1}')
        json_response = unicode(json_response, 'utf-8', errors='ignore')
        jsonobject = simplejson.loads(json_response)
        if jsonobject['result'].has_key('movies'):
            for item in jsonobject['result']['movies']:
                Medialist.append({'movieid': item.get('movieid',''),
                                  'imdbnumber': item.get('imdbnumber',''),
                                  'name': item.get('label',''),
                                  'year': item.get('year',''),
                                  'rating': item.get('rating',''),
                                  'mpaa': item.get('mpaa',''),
                                  'votes': item.get('votes',''),
                                  'file': item.get('file',''),
                                  'path': media_path(item.get('file','')),
                                  'trailer': item.get('trailer','')})

    elif media_type == 'musicvideo':
        json_response = xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "VideoLibrary.GetMusicVideos", "params": {"properties": ["file", "artist", "album", "track", "runtime", "year", "genre"], "sort": { "method": "album" } }, "id": 1}')
        json_response = unicode(json_response, 'utf-8', errors='ignore')
        jsonobject = simplejson.loads(json_response)
        if jsonobject['result'].has_key('musicvideos'):
            for item in jsonobject['result']['musicvideos']:
                Medialist.append({'id': '',
                                  'movieid': item.get('musicvideoid',''),
                                  'name': item.get('label',''),
                                  'artist': item.get('artist',''),
                                  'album': item.get('album',''),
                                  'track': item.get('track',''),
                                  'runtime': item.get('runtime',''),
                                  'year': item.get('year',''),
                                  'path': media_path(item.get('file',''))})
    else:
            log('No JSON results found')
    return Medialist

def __set_library(media_info, new_info):
    xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "VideoLibrary.SetMovieDetails", "params": { "movieid": %i, "votes": %s, "rating": %s, "mpaa": %s}, "id": 1 }' %(media_info['movieid'],new_info['votes'], new_info['rating'], new_info['mpaa']))
    
def media_path(path):
    # Check for stacked movies
    try:
        path = os.path.split(path)[0].rsplit(' , ', 1)[1].replace(",,",",")
    except:
        path = os.path.split(path)[0]
    # Fixes problems with rared movies and multipath
    if path.startswith("rar://"):
        path = [os.path.split(urllib.url2pathname(path.replace("rar://","")))[0]]
    elif path.startswith("multipath://"):
        temp_path = path.replace("multipath://","").split('%2f/')
        path = []
        for item in temp_path:
            path.append(urllib.url2pathname(item))
        print path
    else:
        path = [path]
    return path

resources/lib/provider/imdb.py
#import modules
import xbmc
import sys

### import libraries
from resources.lib.script_exceptions import NoFanartError
from resources.lib.utils import *
from operator import itemgetter
from xml.parsers.expat import ExpatError

### get addon info
__localize__    = ( sys.modules[ "__main__" ].__localize__ )
__addon__    = ( sys.modules[ "__main__" ].__addon__ )


def get_ratings(media_id):
    url = 'http://www.imdbapi.com/?i=%s'
    api_data = get_data(url%(media_id), 'json')
    data = []
    if api_data == "Empty" or not api_data:
        return data
    else:
        # Get fanart
        try:
            data = {'rating': api_data.get('imdbRating',''),
                    'votes': api_data.get('imdbVotes',''),
                    'mpaa': api_data.get('Rated','').upper()}
            if data['rating'] == 'N/A':
                data['rating'] = ''
        except Exception, e:
            log( 'Problem report: %s' %str( e ), xbmc.LOGNOTICE )
        if data == []:
            log('No data for: %s' %media_id)
        else:
            return data

resources/lib/provider/tmdb.py
#import modules
import xbmc
import sys

### import libraries
from resources.lib.script_exceptions import NoFanartError
from resources.lib.utils import *
from operator import itemgetter

### get addon info
__localize__    = ( sys.modules[ "__main__" ].__localize__ )
__addon__    = ( sys.modules[ "__main__" ].__addon__ )

API_KEY = '4be68d7eab1fbd1b6fd8a3b80a65a95e'

def get_ratings(media_id):
    url = 'http://api.themoviedb.org/3/movie/%s?api_key=%s'
    api_data = get_data(url%(media_id, API_KEY), 'json')
    data = []
    if api_data == "Empty" or not api_data:
        return data
    else:
        # Get fanart
        try:
            data = {'rating': api_data.get('vote_average',''),
                    'votes': api_data.get('vote_count','')}
        except Exception, e:
            log( 'Problem report: %s' %str( e ), xbmc.LOGNOTICE )
        if data == []:
            log('No data for: %s' %media_id)
        else:
            return data

def get_releases(media_id):
    url = 'http://api.themoviedb.org/3/movie/%s/releases?api_key=%s'
    api_data = get_data(url%(media_id, API_KEY), 'json')
    data = []
    if api_data == "Empty" or not api_data:
        return data
    else:
        # Got through the list
        for item in api_data['countries']:
            #Check which one matches
            if item.get('iso_3166_1') == __addon__.getSetting("movie_cert_lang_tmdb").upper():
                # strip the date to only the year
                year, month, day = item.get('release_date','').split('-')
                data = {'mpaa': item.get('certification').upper(),
                        'iso_3166_1': item.get('iso_3166_1').upper(),
                        'year': year}
        if data == []:
            log('No data for: %s' %media_id)
        else:
            return data


def _search_movie(medianame,year=''):
    medianame = normalize_string(medianame)
    log('TMDB API search criteria: Title[''%s''] | Year[''%s'']' % (medianame,year) )
    illegal_char = ' -<>:"/\|?*%'
    for char in illegal_char:
        medianame = medianame.replace( char , '+' ).replace( '++', '+' ).replace( '+++', '+' )

    search_url = 'http://api.themoviedb.org/3/search/movie?query=%s+%s&api_key=%s' %( medianame, year, API_KEY )
    tmdb_id = ''
    log('TMDB API search:   %s ' % search_url)
    try:
        data = get_data(search_url, 'json')
        if data == "Empty":
            tmdb_id = ''
        else:
            for item in data['results']:
                if item['id']:
                    tmdb_id = item['id']
                    break
    except Exception, e:
        log( str( e ), xbmc.LOGERROR )
    if tmdb_id == '':
        log('TMDB API search found no ID')
    else:
        log('TMDB API search found ID: %s' %tmdb_id)
    return tmdb_id

default.py
#import modules
import re
import os
import sys
import xbmc
import xbmcaddon
import xbmcgui
import time

### get addon info
__addon__       = xbmcaddon.Addon(id='script.library.info.update')
__addonid__     = __addon__.getAddonInfo('id')
__addonname__   = __addon__.getAddonInfo('name')
__author__      = __addon__.getAddonInfo('author')
__version__     = __addon__.getAddonInfo('version')
__addonpath__   = __addon__.getAddonInfo('path')
__addonprofile__= xbmc.translatePath(__addon__.getAddonInfo('profile')).decode('utf-8')
__icon__        = __addon__.getAddonInfo('icon')
__localize__    = __addon__.getLocalizedString

from resources.lib.utils import *
from resources.lib.media_setup import __get_library as get_library
from resources.lib.media_setup import __set_library as set_library
from resources.lib.provider import tmdb
from resources.lib.provider import imdb
from resources.lib.script_exceptions import *
from xml.parsers.expat import ExpatError
from resources.lib.fileops import fileops

class Main:

    def __init__(self):
        if dialog_msg('yesno',
                      line1 = __localize__(32015),
                      line2 = __localize__(32016),
                      background = False,
                      nolabel = __localize__(32014),
                      yeslabel = __localize__(32013)):
            try:
                # Creates temp folder
                fileops()
            except CreateDirectoryError, e:
                log('Could not create directory: %s' % str(e))
                return False
            else:
                if __addon__.getSetting("movie_enable") == 'true':
                    media_list = get_medialist('movie')
                    if media_list:
                        update_movie(media_list)
            dialog_msg("close")


### retrieve list from library
def get_medialist(media_type):
    return get_library(media_type)


def update_movie(movie_list):
    processeditems = 0
    totalitems = len(movie_list)
    dialog_msg('create', line1 = 'Library info update', background = False)
    # Walk through media list
    for item in movie_list:
        if xbmc.abortRequested or dialog_msg('iscanceled'):
            log('XBMC abort requested, aborting')
            break
        dialog_msg('update',
                   percentage = int(float(processeditems) / float(totalitems) * 100.0),
                   line1 = '%s %s / %s' %(__localize__(32017), processeditems, totalitems),
                   line2 = item['name'],
                   background = False)
        update_media = False
        new_info = {'imdbnumber': 'null',
            'title': 'null',
            'playcount': 'null',
            'runtime': 'null',
            'director': 'null',
            'studio': 'null',
            'year': 'null',
            'plot': 'null',
            'genre': 'null',
            'rating': 'null',
            'mpaa': '""',
            'imdbnumber': 'null',
            'votes': 'null',
            'lastplayed': 'null',
            'originaltitle': 'null',
            'trailer': 'null',
            'tagline': 'null',
            'plotoutline': 'null',
            'writer': 'null',
            'country': 'null',
            'top250': 'null',
            'sorttitle': 'null',
            'set': 'null',
            'showlink': 'null'}
        #Check for IMDB id
        if item['imdbnumber'].startswith('tt'):
            log('------------')
            log(item['name'])
            # Handle themoviedb.org
            if __addon__.getSetting(id="movie_rating") == 'themoviedb.org':
                movie_data = ''
                errmsg = ''
                try:
                    movie_data = tmdb.get_ratings(item['imdbnumber'])
                except HTTP404Error, e:
                    errmsg = '404: File not found'
                    data_result = 'skipping'
                except HTTP503Error, e:
                    errmsg = '503: API Limit Exceeded'
                    data_result = 'retrying'
                except ItemNotFoundError, e:
                    errmsg = '%s not found' % item['name']
                    data_result = 'skipping'
                except HTTPTimeout, e:
                    errmsg = 'Timed out'
                    data_result = 'skipping'
                except DownloadError, e:
                    errmsg = 'Possible network error: %s' % str(e)
                    data_result = 'skipping'
                else:
                    pass
                if errmsg:
                    log(errmsg)
                if movie_data:
                    if item['votes'] != movie_data['votes']:
                        log('new votes: %s, old votes: %s' %(movie_data['votes'], item['votes']))
                        log('new rating: %s, old rating: %s' %(movie_data['rating'], item['rating']))
                        new_info['rating'] = float(movie_data['rating'])
                        new_info['votes'] = '"%s"' %movie_data['votes']
                        update_media = True
            
            if __addon__.getSetting(id="movie_cert") == 'themoviedb.org':
                movie_data = ''
                errmsg = ''
                try:
                    movie_data = tmdb.get_releases(item['imdbnumber'])
                except HTTP404Error, e:
                    errmsg = '404: File not found'
                    data_result = 'skipping'
                except HTTP503Error, e:
                    errmsg = '503: API Limit Exceeded'
                    data_result = 'retrying'
                except ItemNotFoundError, e:
                    errmsg = '%s not found' % item['name']
                    data_result = 'skipping'
                except HTTPTimeout, e:
                    errmsg = 'Timed out'
                    data_result = 'skipping'
                except DownloadError, e:
                    errmsg = 'Possible network error: %s' % str(e)
                    data_result = 'skipping'
                else:
                    pass
                if errmsg:
                    log(errmsg)
                if movie_data:
                    if movie_data['mpaa']:
                        if __addon__.getSetting("movie_cert_nota") == '0':
                            new_mpaa = '"Rated %s"' %movie_data['mpaa']
                        elif __addon__.getSetting("movie_cert_nota") == '1':
                            new_mpaa = '"%s_%s"' %(movie_data['iso_3166_1'], movie_data['mpaa'])
                        else:
                            new_mpaa = '"%s:%s"' %(movie_data['iso_3166_1'], movie_data['mpaa'])
                        if not item['mpaa'] in new_mpaa:
                            log('new mpaa: %s, old mpaa: %s' %(new_mpaa, item['mpaa']))
                            new_info['mpaa'] = new_mpaa
                            update_media = True
            
            if (__addon__.getSetting(id="movie_cert") or __addon__.getSetting(id="movie_cert")) == 'IMDb':
                movie_data = ''
                errmsg = ''
                try:
                    movie_data = imdb.get_ratings(item['imdbnumber'])
                except HTTP404Error, e:
                    errmsg = '404: File not found'
                    data_result = 'skipping'
                except HTTP503Error, e:
                    errmsg = '503: API Limit Exceeded'
                    data_result = 'retrying'
                except ItemNotFoundError, e:
                    errmsg = '%s not found' % item['name']
                    data_result = 'skipping'
                except HTTPTimeout, e:
                    errmsg = 'Timed out'
                    data_result = 'skipping'
                except DownloadError, e:
                    errmsg = 'Possible network error: %s' % str(e)
                    data_result = 'skipping'
                else:
                    pass
                if errmsg:
                    log(errmsg)
                if movie_data:
                    if __addon__.getSetting(id="movie_cert") == 'IMDb' and movie_data['mpaa']:
                        if __addon__.getSetting("movie_cert_nota") == '0':
                            new_mpaa = '"Rated %s"' %movie_data['mpaa']
                        elif __addon__.getSetting("movie_cert_nota") == '1':
                            new_mpaa = '"%s/%s"' %('US', movie_data['mpaa'])
                        else:
                            new_mpaa = '"%s:%s"' %('US', movie_data['mpaa'])
                        if not item['mpaa'] in new_mpaa:
                            log('new mpaa: %s, old mpaa: %s' %(new_mpaa, item['mpaa']))
                            new_info['mpaa'] = new_mpaa
                            update_media = True
                    if __addon__.getSetting(id="movie_rating") == 'IMDb' and movie_data['rating']:
                        if item['votes'] != movie_data['votes']:
                            log('new votes: %s, old votes: %s' %(movie_data['votes'], item['votes']))
                            log('new rating: %s, old rating: %s' %(movie_data['rating'], item['rating']))
                            new_info['rating'] = float(movie_data['rating'])
                            new_info['votes'] = '"%s"' %movie_data['votes']
                            update_media = True

            log('update: %s' %update_media)
            if update_media:
                set_library(item, new_info)
                pass
        processeditems += 1


### Start of script
if (__name__ == '__main__'):
    log('######## Library info update: Initializing...............................', xbmc.LOGNOTICE)
    log('## Add-on ID   = %s' % str(__addonid__), xbmc.LOGNOTICE)
    log('## Add-on Name = %s' % str(__addonname__), xbmc.LOGNOTICE)
    log('## Authors     = %s' % str(__author__), xbmc.LOGNOTICE)
    log('## Version     = %s' % str(__version__), xbmc.LOGNOTICE)
    Main()
    log('## Script stopped or ended', xbmc.LOGNOTICE)